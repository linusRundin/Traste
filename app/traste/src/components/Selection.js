import React from 'react';
import PropTypes from 'prop-types';
import {TextField, MenuItem} from '@mui/material';

/**
 * Custom TextField for displaying a selectionbox with different alternatives.
 * @param {*} label Name of field.
 * @param {*} value Starting value.
 * @param {*} onChange Handler.
 * @param {*} error What will be shown for the error.
 * @param {*} data A list of data to be displayed.
 * @return {TextField} Rendred field.
 */
function Selection({label, value, onChange, error, data}) {
  return (
    <TextField
      select
      label={label}
      value={value}
      fullWidth
      onChange={onChange}
      sx={{
        marginTop: '15px',
        backgroundColor: 'rgba(255,255,255,0.3)',
      }}
      inputProps={{'data-testid': 'selectionfield'}}
      error={!!error}
    >
      {
        /** Placeholder if thereÂ´s no data available */
        data !== undefined ? (
          data.map((option) => (
            <MenuItem
              key={option.id} value={option.label}>
              {option.label}
            </MenuItem>
          ))
        ) : (
          <MenuItem key={0}>No data</MenuItem>
        )
      }
    </TextField>
  );
}


Selection.defaultProps = {
  error: false,
};

Selection.propTypes = {
  label: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.any.isRequired,
  data: PropTypes.array,
};

export default Selection;
