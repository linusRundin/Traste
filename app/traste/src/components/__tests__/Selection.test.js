import {cleanup} from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import Selection from '../Selection';
import React from 'react';
import {configure, shallow} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import {MenuItem} from '@mui/material';

configure({adapter: new Adapter()});

afterEach(() => {
  cleanup();
});

const propsData = {
  title: 'Bin Size',
  name: 'BinSize',
  label: 'Selection',
  value: '',
  error: 'error',
};
const data = [
  {
    id: '0',
    label: '5',
  },
  {
    id: '1',
    label: '10',
  },
  {
    id: '2',
    label: '15',
  },
  {
    id: '3',
    label: '20',
  },
];

test('Check that MenuItem, in the Selection component, has' +
'text No data because data props is null', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<Selection
    title={propsData.title}
    name={propsData.name}
    {...propsData}
    onChange={onChange}
  />);

  expect(wrapper.find(MenuItem)).toHaveLength(1);
  expect(wrapper.find(MenuItem).text()).toBe('No data');
});

test('Check so that Selection component contains 4 values', () => {
  const onChange = jest.fn();
  const wrapper = shallow(<Selection
    title={propsData.title}
    name={propsData.name}
    data={data}
    {...propsData}
    onChange={onChange}
  />);

  expect(wrapper.find(MenuItem)).toHaveLength(4);
});


test('Matches snapshot', () => {
  const onChange = jest.fn();

  const tree = renderer.create(<Selection
    title={propsData.title}
    name={propsData.name}
    data={data}
    {...propsData}
    onChange={onChange}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});
