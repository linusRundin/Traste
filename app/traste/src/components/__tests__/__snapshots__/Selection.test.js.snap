// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1uxunh6-MuiFormControl-root-MuiTextField-root"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={false}
  >
    Selection
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
    variant="outlined"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="notranslate"
      >
        â€‹
      </span>
    </div>
    <input
      aria-hidden={true}
      autoFocus={false}
      className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      data-testid="selectionfield"
      disabled={false}
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value=""
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-173wfxe"
      >
        <span>
          Selection
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
